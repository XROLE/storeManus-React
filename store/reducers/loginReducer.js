import axios from 'axios';
import {
  LOGIN_SUCCESS,
  LOGIN_PENDING,
  LOGIN_ERROR,
  ADDING_ATTENDANT,
  ADDED_ATTENDANT,
  ADD_ATTENDANT_ERROR,
} from '../actions/actionTypes';

const aMail = 'xrolediamond@gmail.com';
const baseUrl = 'https://storemanus.herokuapp.com/api/v1';
const adminUrl = `${baseUrl}/admin/auth/signin`;
const attendantsUrl = `${baseUrl}/attendants/auth/signin`;

const headers = {
  headers: {
    Accept: 'application/json, text/plain, */*',
    'x-access-token': `Bearer ${localStorage.getItem('storeManus_token')}`,
    'Content-type': 'application/json',
  },
};

const initialState = {
  token: '',
  user: '',
  pending: false,
  error: '',
};


export default function (state = initialState, action) {
  switch (action.type) {
    case LOGIN_ERROR: {
      return {
        ...state,
        error: action.error,
        pending: false,
      };
    }
    case LOGIN_SUCCESS: {
      return {
        ...state,
        token: action.payload.Token,
        user: action.payload.User,
        pending: false,
      };
    }
    case LOGIN_PENDING: {
      return {
        ...state,
        error: null,
        pending: true,
      };
    }
    default:
      return state;
  }
}


export function loginInProgress() {
  return { type: LOGIN_PENDING };
}

export function addingAttendantInProgress() {
  return { type: ADDING_ATTENDANT };
}

export function loginUser(loginDetails) {
  let url;

  return (dispatch) => {
    const { email } = loginDetails;
    if (email === aMail) {
      url = adminUrl;
    } else {
      url = attendantsUrl;
    }
    dispatch(loginInProgress());
    axios
      .post(url, loginDetails, headers)
      .then((res) => {
        const { Token, User } = res.data;
        localStorage.setItem('accessToken', Token);
        dispatch({ type: LOGIN_SUCCESS, payload: { Token, User } });
      }).catch((err) => {
        const error = err.response.data.Message;
        dispatch({ type: LOGIN_ERROR, error });
      });
  };
}

export function addAttendant(attendantDetails) {
  const url = `${baseUrl}/attendants/auth/register`;

  return (dispatch) => {
    dispatch(addingAttendantInProgress());
    axios
      .post(url, attendantDetails, headers)
      .then((res) => {
        const message = `Proceed to update your account password = ${res.data.password}`;
        dispatch({ type: ADDED_ATTENDANT, payload: { message } });
        alert(message);
      }).catch((err) => {
        const error = err.response.data.Message;
        console.log(err);
        dispatch({ type: ADD_ATTENDANT_ERROR, error });
      });
  };
}
