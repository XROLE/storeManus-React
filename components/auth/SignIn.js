import React, { Component } from 'react';
import { ToastContainer, toast } from 'react-toastify';
import { Provider, connect } from 'react-redux';
import { Redirect } from 'react-router-dom';
import PropTypes from 'prop-types';
import Spinner from 'react-spinkit';
import Footer from '../footer/footer';
import { loginUser } from '../../store/reducers/loginReducer';
import Navbar from '../navbar/Navbar';
import store from '../../store/index';
import 'react-toastify/dist/ReactToastify.css';

class SignIn extends Component {
  constructor(props) {
    super(props);
    this.state = {
      email: '',
      password: '',
      shouldShowError: true,
    };
  }

  onChange(e) {
    return this.setState({
      [e.target.name]: e.target.value.trim(),
    });
  }

  handleSubmit(e) {
    e.preventDefault();

    const loginDetails = this.state;
    const { login } = this.props;
    this.setState({
      shouldShowError: true,
    });
    login(loginUser(loginDetails));
  }

  notify(error) {
    toast.error(error);
  }

  render() {
    const { email, password, shouldShowError } = this.state;
    const { pending, loginError, token } = this.props;

    if (loginError && shouldShowError) {
      this.notify(loginError);
      this.setState({
        shouldShowError: false,
      });
    }

    if (token && email === 'xrolediamond@gmail.com') {
      localStorage.removeItem('storeManus_token');
      localStorage.setItem('storeManus_token', token);
      return <Redirect to="/adminDashboard" />;
    } if (token) {
      return <Redirect to="/attendantsDashboard" />;
    }
    return (
      <Provider store={store}>
        <Navbar />
        <div className="signupDiv" style={{ marginTop: '180px' }}>
          <form action="" className="signup-form">
            <div>
              <input
                type="email"
                name="email"
                id="signInEmail"
                placeholder="Email"
                onChange={e => this.onChange(e)}
                value={email}
              />
            </div>
            <div>
              <input
                type="password"
                name="password"
                id="signInPassword"
                placeholder="Password"
                onChange={e => this.onChange(e)}
                value={password}
              />
            </div>
            <div className="signup-button-div">
              <button
                className="link signin-button"
                type="submit"
                onClick={e => this.handleSubmit(e)}
              >
                SIGN IN
              </button>
            </div>
            { pending === true && (
            <Spinner name="circle" className="spinner" id="reactLoader" />
            )}
          </form>
        </div>
        <Footer />
        <ToastContainer />
      </Provider>
    );
  }
}
SignIn.propTypes = {
  login: PropTypes.func.isRequired,
  pending: PropTypes.bool.isRequired,
  // eslint-disable-next-line react/require-default-props
  loginError: PropTypes.string,
  token: PropTypes.string.isRequired,
};

const mapStateToProps = state => ({
  token: state.login.token,
  pending: state.login.pending,
  loginError: state.login.error,
});

function mapDispatchToProps(dispatch) {
  return ({
    login: action => dispatch(action),
  });
}

export default connect(mapStateToProps, mapDispatchToProps)(SignIn);
