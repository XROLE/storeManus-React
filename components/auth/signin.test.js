import React from 'react';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { shallow, mount } from 'enzyme';
import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { SignIn } from './SignIn';
import loginReducer from '../../store/reducers/loginReducer';


describe('Login Component', () => {
  it('Should render just the SignIn component', () => {
    const mockLoginfn = jest.fn();
    const token = '';
    const pending = false;
    const component = shallow(
      <SignIn
        login={mockLoginfn}
        token={token}
        pending={pending}
      />,
    );
    expect(component.exists()).toBe(true);
    expect(component).toMatchSnapshot();
  });
});

describe('Login Component', () => {
  const mockLoginfn = jest.fn();
  const token = '';
  const pending = false;
  const wrapper = shallow(<SignIn
    login={mockLoginfn}
    token={token}
    pending={pending}
  />);

  it('should call the mock login function', () => {
    wrapper.setState({ email: '', password: '', shouldShowError: false });
    wrapper.find('form').simulate(
      'submit',
      {
        preventDefault: () => {},
        target: {
          name: {
            email: { value: 'johndoe@joe.com' },
            password: { value: 'johndoe88' },
          },
        },
      },
    );
    expect(mockLoginfn).toHaveBeenCalled();
  });
});


describe('Login Functionality ', () => {
  const login = jest.fn();
  const initialState = {
    login: {
      token: '',
      user: '',
      pending: false,
      error: '',
    },
  };
  const mockStore = configureStore([thunk]);
  const store = mockStore(() => initialState);
  // const dispatchMock = jest.spyOn(store, 'dispatch');
  const pending = false;
  const loginError = '';
  const token = '';
  const success = false;

  const props = {
    profile: {
      login,
      pending,
      loginError,
      token,
      success,
    },
  };

  const wrapper = mount(
    <Provider store={store}>
      <BrowserRouter>
        <SignIn {...props} />
      </BrowserRouter>
    </Provider>,
  );

  //   it('should call the login function', () => {
  //     wrapper.find('#signInEmail').simulate(
  //       'change', {
  //         target: {
  //           name: 'email',
  //           value: 'jade@gmail.com',
  //         },
  //       },
  //     );

  //     wrapper.find('#signInPassword').simulate(
  //       'change', {
  //         target: {
  //           name: 'password', value: 'tedTryCatch',
  //         },
  //       },
  //     );

//     wrapper.find('.signin-button').simulate('click');
//     expect(dispatchMock).toHaveBeenCalled();
//   });
});

it('should setup default state values', () => {
  const state = loginReducer(undefined, {
    type: '@@INIT',
  });
  expect(state).toEqual({
    token: '',
    user: '',
    pending: false,
    error: '',
  });
});

it('should update state when action type is  LOGIN_PENDING', () => {
  const state = loginReducer(undefined, {
    type: 'LOGIN_PENDING',
  });
  expect(state).toEqual({
    ...state,
    pending: true,
  });
});

it('should update state when action type is  LOGIN_SUCCESS', () => {
  const state = loginReducer(undefined, {
    type: 'LOGIN_SUCCESS',
    payload: {
      success: true,
    },
  });
  expect(state).toEqual({
    ...state,
    success: true,
  });
});

it('should update state when action type is  LOGIN_ERROR', () => {
  const state = loginReducer(undefined, {
    type: 'LOGIN_ERROR',
    error: '',
  });
  expect(state).toEqual({
    ...state,
    error: '',
    pending: false,
  });
});
