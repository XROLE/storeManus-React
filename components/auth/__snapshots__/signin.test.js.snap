// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Component Should render just the SignIn component 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Navbar />
  <div
    className="signupDiv"
    style={
      Object {
        "marginTop": "180px",
      }
    }
  >
    <form
      action=""
      className="signup-form"
      onSubmit={[Function]}
    >
      <div>
        <input
          id="signInEmail"
          name="email"
          onChange={[Function]}
          placeholder="Email"
          type="email"
          value=""
        />
      </div>
      <div>
        <input
          id="signInPassword"
          name="password"
          onChange={[Function]}
          placeholder="Password"
          type="password"
          value=""
        />
      </div>
      <div
        className="signup-button-div"
      >
        <button
          className="link signin-button"
          onClick={[Function]}
          type="submit"
        >
          SIGN IN
        </button>
      </div>
    </form>
  </div>
  <Footer />
  <ToastContainer
    autoClose={5000}
    bodyClassName={null}
    className={null}
    closeButton={
      <CloseButton
        ariaLabel="close"
      />
    }
    closeOnClick={true}
    draggable={true}
    draggablePercent={80}
    hideProgressBar={false}
    newestOnTop={false}
    pauseOnFocusLoss={true}
    pauseOnHover={true}
    position="top-right"
    progressClassName={null}
    progressStyle={null}
    rtl={false}
    style={null}
    toastClassName={null}
    transition={[Function]}
  />
</Provider>
`;
